### PR TITLE
{2023.06}[foss/2021a] WRF-dmpar V4.3

### PR DESCRIPTION
**Pull request by TopRichard**

### CODE REVIEW

**Comment by boegel** [comment]
Installation failed in configure step, details below.

Looks like our pre-configure hook for WRF needs work?

<details>

```
== configuring...
== Running pre-configure hook...
== Using custom preconfigopts for WRF:  sed -i 's/Linux x86_64 ppc64le, gfortran/Linux x86_64 aarch64 ppc64le, gfortran/g' arch/configure_new.defaults &&
  >> running interactive command:
	[started at: 2023-06-29 07:47:46]
	[working dir: /cvmfs/pilot.eessi-hpc.org/versions/2023.06/software/linux/aarch64/generic/software/WRF/4.3-foss-2021a-dmpar/WRF-4.3]
	[output logged in /tmp/eb-dxromz72/eb-a8u9yx8o/eb-yad5i647/eb-vayxs19v/eb-p6cdsroq/eb-hoeqxajd/eb-b4whiekr/eb-8z9yjdda/easybuild-run_cmd_qa-3x6bwut9.log]
	sed -i 's/Linux x86_64 ppc64le, gfortran/Linux x86_64 aarch64 ppc64le, gfortran/g' arch/configure_new.defaults &&   ./configure
  >> interactive command completed: exit 2, ran in 00h00m01s
== ... (took 1 secs)
== FAILED: Installation ended unsuccessfully (build directory: /cvmfs/pilot.eessi-hpc.org/versions/2023.06/software/linux/aarch64/generic/software/WRF/4.3-foss-2021a-dmpar): build failed (first 300 chars): cmd " sed -i 's/Linux x86_64 ppc64le, gfortran/Linux x86_64 aarch64 ppc64le, gfortran/g' arch/configure_new.defaults &&   ./configure " exited with exit code 2 and output:
sed: can't read arch/configure_new.defaults: No such file or directory
 (took 36 secs)
```

</details>

**Comment by bedroge** [comment]
> Installation failed in configure step, details below.
> 
> Looks like our pre-configure hook for WRF needs work?
> 
> ```
> == configuring...
> == Running pre-configure hook...
> == Using custom preconfigopts for WRF:  sed -i 's/Linux x86_64 ppc64le, gfortran/Linux x86_64 aarch64 ppc64le, gfortran/g' arch/configure_new.defaults &&
>   >> running interactive command:
> 	[started at: 2023-06-29 07:47:46]
> 	[working dir: /cvmfs/pilot.eessi-hpc.org/versions/2023.06/software/linux/aarch64/generic/software/WRF/4.3-foss-2021a-dmpar/WRF-4.3]
> 	[output logged in /tmp/eb-dxromz72/eb-a8u9yx8o/eb-yad5i647/eb-vayxs19v/eb-p6cdsroq/eb-hoeqxajd/eb-b4whiekr/eb-8z9yjdda/easybuild-run_cmd_qa-3x6bwut9.log]
> 	sed -i 's/Linux x86_64 ppc64le, gfortran/Linux x86_64 aarch64 ppc64le, gfortran/g' arch/configure_new.defaults &&   ./configure
>   >> interactive command completed: exit 2, ran in 00h00m01s
> == ... (took 1 secs)
> == FAILED: Installation ended unsuccessfully (build directory: /cvmfs/pilot.eessi-hpc.org/versions/2023.06/software/linux/aarch64/generic/software/WRF/4.3-foss-2021a-dmpar): build failed (first 300 chars): cmd " sed -i 's/Linux x86_64 ppc64le, gfortran/Linux x86_64 aarch64 ppc64le, gfortran/g' arch/configure_new.defaults &&   ./configure " exited with exit code 2 and output:
> sed: can't read arch/configure_new.defaults: No such file or directory
>  (took 36 secs)
> ```

Just had a quick look at the contents of the source tarball: apparently, version 3.9.x had `arch/configure_{old,new}.defaults`, while version 4.3 only has a `arch/configure.defaults`. The latter already includes the following:
```
#ARCH    Linux i486 i586 i686 armv7l aarch64, gfortran compiler with gcc #serial smpar dmpar dm+sm
```
So I guess the hook is no longer required for newer version, and we should add a version check to it.

**Comment by bedroge** [comment]
> 
> Just had a quick look at the contents of the source tarball: apparently, version 3.9.x had `arch/configure_{old,new}.defaults`, while version 4.3 only has a `arch/configure.defaults`. The latter already includes the following:
> 
> ```
> #ARCH    Linux i486 i586 i686 armv7l aarch64, gfortran compiler with gcc #serial smpar dmpar dm+sm
> ```
> 
> So I guess the hook is no longer required for newer version, and we should add a version check to it.

Checked some more versions: 
 - 3.9 has these `old/new` files
 - 4.0 up to and including 4.2.1 just have `configure.defaults`, but no support for `aarch64` yet
 - 4.2.2 and newer do have `aarch64` listed in `configure.defaults`

So the hook would have to check for these three version ranges and do the corresponding action.

File: eb_hooks.py [hunk 1]
```diff
@@ -251,11 +251,17 @@ def pre_configure_hook_wrf_aarch64(self, *args, **kwargs):
     - patch arch/configure_new.defaults so building WRF with foss toolchain works on aarch64
     """
     if self.name == 'WRF':
+        wrfversion = int((self.version).replace(".",""))
         if get_cpu_architecture() == AARCH64:
             pattern = "Linux x86_64 ppc64le, gfortran"
             repl = "Linux x86_64 aarch64 ppc64le, gfortran"
-            self.cfg.update('preconfigopts', "sed -i 's/%s/%s/g' arch/configure_new.defaults && " % (pattern, repl))
-            print_msg("Using custom preconfigopts for %s: %s", self.name, self.cfg['preconfigopts'])
+            if wrfversion <= 39:
```
Comment (boegel) [inline eb_hooks.py:258]: @TopRichard It's better to do version checks with `LooseVersion`, and not convert to an integer (consider WRF versions like `3.9.1.1`):

```suggestion
            if LooseVersion(self.version) <= LooseVersion('3.9.0'):

**Comment by TopRichard** [comment]
bot: build repo:eessi-2023.06-software arch:x86_64/generic
bot: build repo:eessi-2023.06-software arch:x86_64/intel/haswell
bot: build repo:eessi-2023.06-software arch:x86_64/intel/skylake_avx512
bot: build repo:eessi-2023.06-software arch:x86_64/amd/zen3
bot: build repo:eessi-2023.06-software arch:x86_64/amd/zen2
bot: build repo:eessi-2023.06-software arch:aarch64/neoverse_v1
bot: build repo:eessi-2023.06-software arch:aarch64/neoverse_n1

**Comment by boegel** [comment]
bot: build repo:eessi-2023.06-software arch:aarch64/neoverse_v1

**Comment by boegel** [comment]
@TopRichard Can you re-try the build for aarch64/neoverse-v1 using the update WRF easyblock from https://github.com/easybuilders/easybuild-easyblocks/pull/3006, just to see if that helps at all? May not make a difference...

**Comment by TopRichard** [comment]
bot: build repo:eessi-2023.06-software arch:aarch64/neoverse_v1

**Pull request closed without merging**